### ISSUE TITLE
computation of zeta for stable buoyancy forcing and wind stress

### ISSUE THREAD

**Issue by vanroekel** [CVMix/CVMix-src#56]
As I have been looking through CVMIX, I have an issue with the following code in cvmix_kpp_compute_turbulent_scales_1d.  In it, zeta (sigma * h / L) is computed as follows

```
 if (surf_fric_vel.ne.cvmix_zero) then
      do kw=1,n_sigma
        ! compute scales at sigma if sigma < surf_layer_ext, otherwise compute
        ! at surf_layer_ext
        zeta(kw) = min(surf_layer_ext, sigma_coord(kw)) * OBL_depth *         &
                   surf_buoy_force*vonkar/(surf_fric_vel**3)
      end do
```

I am wondering why the convective case limitation on sigma is included for stable buoyancy forcing as well?  Large et al. (1994), Troen and Mahrt (1986), and the CVMIX description document assume that the similarity functions for the surface layer extend from 0 < sigma < 1

I have run a test case with positive buoyancy forcing and a constant wind stress in a SCM framework.  Below is a plot of temperature profiles through time, with the kpp diagnosed boundary layer depth in black.

![mpas_shear_fix_wind_stable_buoy](https://cloud.githubusercontent.com/assets/13662545/10475034/c70f70aa-71fb-11e5-9389-adc2d7408a41.png)

The corresponding result from a similarly forced LES is below (the line is a diagnosed mixed layer depth from a temperature threshold criterion)

![les_wind_stable_buoy](https://cloud.githubusercontent.com/assets/13662545/10475041/e77423c2-71fb-11e5-8d98-d9849baa7cc1.png)

I have then run a test that changed the CVMIX code above to

```
    if (surf_fric_vel.ne.cvmix_zero) then
      do kw=1,n_sigma
        ! compute scales at sigma if sigma < surf_layer_ext, otherwise compute
        ! at surf_layer_ext
        if(surf_buoy_force .ge. cvmix_zero) then
                zeta(kw) = sigma_coord(kw) * OBL_depth *                      &
                        surf_buoy_force*vonkar/(surf_fric_vel**3)
        else
                zeta(kw) = min(surf_layer_ext, sigma_coord(kw)) * OBL_depth *         &
                   surf_buoy_force*vonkar/(surf_fric_vel**3)
        endif
      end do
```

with this change, the temperature profile plot becomes

![cvmix_turb_scale_fix_wind_stable_buoy](https://cloud.githubusercontent.com/assets/13662545/10475058/1980eefe-71fc-11e5-8001-a80b8594c10b.png)

which is closer to the LES result.

It seems that the in stable conditions, sigma should not be confined less than the surface layer extent.

**Comment by mnlevy1981** [comment]
This change is documented in the 2006 Danabasoglu, et al., paper [Diurnal Coupling in the Tropical Oceans of CCSM3](http://journals.ametsoc.org/doi/abs/10.1175/JCLI3739.1) (found in Journal of Climate). From Appendix A (page 2361):

> The turbulent velocity scale, ws, usually equals the friction velocity at the surface and increases in unstable forcing over the uppermost 10% of the boundary layer, then is held constant below in order to avoid excessively large values as depth increases. For consistency, a similar constraint is now applied in stable forcing that limits the decrease in ws with depth, and hence increases Vt.

**Comment by StephenGriffies** [comment]
Luke,

I am having a tough time seeing a significant difference between your two
KPP runs.  It would help to have the three figures together so that I can
better compare.  But from my initial look, it is tough to say that your
modified KPP run is closer to LES.

 Perhaps I am not seeing something that you are focusing on...?

Steve



On Wed, Oct 14, 2015 at 11:24 AM, Michael Levy <notifications@github.com>
wrote:

> This change is documented in the 2006 Danabasoglu, et al., paper Diurnal
> Coupling in the Tropical Oceans of CCSM3
> <http://journals.ametsoc.org/doi/abs/10.1175/JCLI3739.1> (found in
> Journal of Climate). From Appendix A (page 2361):
>
> The turbulent velocity scale, ws, usually equals the friction velocity at
> the surface and increases in unstable forcing over the uppermost 10% of the
> boundary layer, then is held constant below in order to avoid excessively
> large values as depth increases. For consistency, a similar constraint is
> now applied in stable forcing that limits the decrease in ws with depth,
> and hence increases Vt.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/CVMix/CVMix-src/issues/56#issuecomment-148083738>.
>



-- 
Dr. Stephen M. Griffies
NOAA Geophysical Fluid Dynamics Lab
201 Forrestal Road
Princeton, NJ 08542
USA

**Comment by vanroekel** [comment]
Steve,
  It is certainly not a massive difference, but it is there.  I've annotated the images from the two runs and placed them closer together below.  Perhaps this will help.  I've added two vertical lines at the same times (1.5 days and 3.0 days).  What I'm noticing is that at day 1.5, the reddish contour has already outcropped/reached the surface in the CVMIX formulation from Danabasoglu et al 2006.  When the turbulent scale is not limited, the temperature stays warmer longer.  This is seen in the LES.  If you examine day 1.5, the temperature structure matches better with my test case (altering CVMIX).  If you then look at day 3, the dark red temperature surface has nearly reached the surface in the current CVMIX form, but has not in the test case.  Again, this result is closer to what is observed in LES.  As alluded to in Danabasoglu et al 2006, the artificial limitation of sigma deepens the boundary layer depth minimally (approximately 0.5 meters difference in the two cases by the end of 3.5 days)

Run with sigma limitation (current cvmix)

![mpas_shear_fix_wind_stable_buoy](https://cloud.githubusercontent.com/assets/13662545/10492837/48455918-726b-11e5-89be-b09385f992e6.png)

without the limitation on sigma

![cvmix_turb_scale_fix_wind_stable_buoy](https://cloud.githubusercontent.com/assets/13662545/10492914/99bb63f0-726b-11e5-97df-93e707c83b4b.png)

Finally LES

![les_wind_stable_buoy](https://cloud.githubusercontent.com/assets/13662545/10492971/d7aabe54-726b-11e5-8bac-1168966c37dd.png)

In essence, it appears to me that the temperature structure observed with the cvmix changed is closer to the LES result (horizontally averaged temperature profiles).  If this is still clear as mud to others besides myself, I can go back and form plots of theta_original_cvmix - theta_LES and theta_changed_cvmix - theta_LES.

In general, however, I'm not totally comfortable with, or perhaps not understanding, the reasoning for making the change outlined Danabasoglu et al 2006.  By my reading, it seems that the turbulent velocity scaling was changed to make CCSM3 boundary layer depths deeper in the tropical simulations shown in the paper (closer to observations).  But, there is really no physical rationale behind this change.  Then again, there really is no physical rationale to changing back to the original LMD94 formulation.  One of my current tasks is to develop an analysis to help visualize the turbulent velocity profile in the boundary layer in my corresponding LES result.  I will post this figure here as soon as I get it made.

**Comment by StephenGriffies** [comment]
Hi,

I now see your point.  Thanks for the clarification.

One suggestion is to code an option to facilitate further testing of your
modified approach.  Can you fork a branch that provides the code mod that
allows for this change to be toggled, with the default set to the current
settings?  Please document your code mod so that folks can know what is at
question.  Be precise, in that you should point to the Danabasoglu et al
(2006) equations/discussion that you are questioning.

Please be sure to allow for present answers to go unchanged, as that will
facilitate easier testing by others.

When you are happy with your fork, then let others know so that they can
test the change.  If others concur, we may wish to incorporate this to the
main branch.  But I defer to the NCAR folks to make that decision...

Thanks,
  Steve





On Wed, Oct 14, 2015 at 2:10 PM, Luke Van Roekel <notifications@github.com>
wrote:

> Steve,
> It is certainly not a massive difference, but it is there. I've annotated
> the images from the two runs and placed them closer together below. Perhaps
> this will help. I've added two vertical lines at the same times (1.5 days
> and 3.0 days). What I'm noticing is that at day 1.5, the reddish contour
> has already outcropped/reached the surface in the CVMIX formulation from
> Danabasoglu et al 2006. When the turbulent scale is not limited, the
> temperature stays warmer longer. This is seen in the LES. If you examine
> day 1.5, the temperature structure matches better with my test case
> (altering CVMIX). If you then look at day 3, the dark red temperature
> surface has nearly reached the surface in the current CVMIX form, but has
> not in the test case. Again, this result is closer to what is observed in
> LES. As alluded to in Danabasoglu et al 2006, the artificial limitation of
> sigma deepens the boundary layer depth minimally (approximately 0.5 meters
> difference in the two cases by the end of 3.5 days)
>
> Run with sigma limitation (current cvmix)
>
> [image: mpas_shear_fix_wind_stable_buoy]
> <https://cloud.githubusercontent.com/assets/13662545/10492837/48455918-726b-11e5-89be-b09385f992e6.png>
>
> without the limitation on sigma
>
> [image: cvmix_turb_scale_fix_wind_stable_buoy]
> <https://cloud.githubusercontent.com/assets/13662545/10492914/99bb63f0-726b-11e5-97df-93e707c83b4b.png>
>
> Finally LES
>
> [image: les_wind_stable_buoy]
> <https://cloud.githubusercontent.com/assets/13662545/10492971/d7aabe54-726b-11e5-8bac-1168966c37dd.png>
>
> In essence, it appears to me that the temperature structure observed with
> the cvmix changed is closer to the LES result (horizontally averaged
> temperature profiles). If this is still clear as mud to others besides
> myself, I can go back and form plots of theta_original_cvmix - theta_LES
> and theta_changed_cvmix - theta_LES.
>
> In general, however, I'm not totally comfortable with, or perhaps not
> understanding, the reasoning for making the change outlined Danabasoglu et
> al 2006. By my reading, it seems that the turbulent velocity scaling was
> changed to make CCSM3 boundary layer depths deeper in the tropical
> simulations shown in the paper (closer to observations). But, there is
> really no physical rationale behind this change. Then again, there really
> is no physical rationale to changing back to the original LMD94
> formulation. One of my current tasks is to develop an analysis to help
> visualize the turbulent velocity profile in the boundary layer in my
> corresponding LES result. I will post this figure here as soon as I get it
> made.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/CVMix/CVMix-src/issues/56#issuecomment-148140990>.
>



-- 
Dr. Stephen M. Griffies
NOAA Geophysical Fluid Dynamics Lab
201 Forrestal Road
Princeton, NJ 08542
USA

**Comment by vanroekel** [comment]
Steve (et al.),
  I have created a fork and put in a pull request (https://github.com/CVMix/CVMix-src/pull/57).  I was not sure if this was where you wanted me to reference the fork for testing.  The relevant text and equations from Large et al 1994 and Danabasoglu et al 2006 are included in the PR description.  I will also continue testing this branch in my SCM version of MPAS-O relative to LES.

Regards,
Luke

**Comment by mnlevy1981** [comment]
Fixed via #62 on March 7th
